[Raisin Bread] instruments and data

ridz1208 opened the pull request:
Remake of https://github.com/aces/Loris/pull/3788 and https://github.com/aces/Loris/pull/3600

this PR does much more than it should but it is here to see if travis fails when the data is modified

SQL Changes: RBdata.sql
 - resolved discrepency between subprojects being `11,12,13,14` in some places and `1,2,3,4` in other places
 - made sure every instrument has a table in sql and metainformation. removed all entries in battery that have no instrument, removed any instrument that has no table, removed any table that has no instrument.
 - broken down to multiple files easier to modify when necessary in PRs

Instrument changes:
- added meta files for each example instrument
- updated instruments to new version when possible
- added sql create tables
- moved testing instruments to test folder

johnsaigle commented:
Travis error: `ERROR 1054 (42S22) at line 1353: Unknown column 'PendingStaging' in 'field list'`

kongtiaowang commented on Dockerfile.test.db:
```diff
@@ -7,26 +7,31 @@ COPY SQL/0000-00-01-Permission.sql /0000-00-01-Permission.sql
-    radiology_review.sql > /docker-entrypoint-initdb.d/0000-compiled.sql 
+    mri_parameter_form.sql \
+    radiology_review.sql \
+    RBdata_meta_prerequesites.sql \
```
RBdata_consent_info_history.sql  /

ridz1208 replied:
Consent info history was removed in 20.0.0

I will update the docker

ridz1208 commented:
@kongtiaowang I made the changes, its still failing

kongtiaowang commented:
@ridz1208 I need to make a PR for fixing the Travis and docker can't load config.xml and read the ajax file properly first, then we need to rebase this PR. I will fix this PR later.

ridz1208 commented:
TODO:

I need to make sure the visit_projec_subproject is properly populated

kongtiaowang commented:
https://github.com/aces/Loris/pull/4145

kongtiaowang commented:
test/unittests/VisitTest.php

kongtiaowang commented:
@ridz1208 
https://github.com/aces/Loris/pull/4155/
Two files need to be modified.  
Database.class.inc  and test/unittests/VisitTest.php

paiva commented on docs/instruments/NDB_BVL_Instrument_TEMPLATE.class.inc:
```diff
@@ -1,89 +1,130 @@
-        //Defines the call back function for HTML Quickform to use when validating the form.
-        $this->form->addFormRule(array(&$this,'XINValidate'));
+        // Defines the call back function for HTML Quickform to use in validation
+        $this->form->addFormRule(array(&$this, 'XINValidate'));
```
Shouldn't it be `([&$this, 'XINValidate'])` instead of `array(...)`?

johnsaigle replied:
No our standard for declaring arrays in PHp is using `array()` instead of `[]`.

ridz1208 commented:
TODO:

- [ ] Fix data in Project_rel table
- [ ] fill data in visit, visit_subproject_project tables
- [ ] update to version 20.1 + major changes

kongtiaowang commented:
@ridz1208 
 Cannot truncate a table referenced in a foreign key constraint (`LorisTest`.`Config`, CONSTRAINT `fk_Config_1` FOREIGN KEY (`ConfigID`) REFERENCES `LorisTest`.`ConfigSettings` (`ID`))

kongtiaowang commented:
test/RBfiles/RB_candidate.sql   =>  Gender to Sex

kongtiaowang commented:
https://github.com/kongtiaowang/Loris/pull/208

ridz1208 commented on SQL/0000-00-00-schema.sql:
```diff
@@ -359,7 +359,7 @@ CREATE TABLE `test_subgroups` (
 
 
-INSERT INTO test_subgroups (Subgroup_name) VALUES ('Instruments');
+INSERT INTO test_subgroups (Subgroup_name) VALUES ('Instruments'),('Imaging');
```
@driusan I added this to the schema becaue some of our "template" instruments assume it exists...

should I add it here or remove it from the instruments ?

driusan replied:
I think it makes sense to include it here

ridz1208 commented:
@kongtiaowang I cleaned up the tables quite a bit and sourced all the new patches and clean-up patches.

I also changed the directory of theRB_files to put them under SQL

can you let me know why it's still failing now ?

kongtiaowang commented:
ERROR 1146 (42S02) at line 81008: Table 'LorisTest.physiological_archive' doesn't exist

kongtiaowang commented:
ERROR 1054 (42S22) at line 26892: Unknown column 'CenterID' in 'field list'

kongtiaowang commented:
RB_candidate.sql, switch CenterID to RegistrationCenterID

ridz1208 commented:
@kongtiaowang 

> ERROR 1054 (42S22) at line 27181: Unknown column 'CenterID' in 'field list'

Any way to know which table is generating that error ? I compared all the files in `RB_files` with the schema files and the columns CenterID match...

driusan commented on SQL/RB_files/RB_Config.sql:
```diff
@@ -0,0 +1,90 @@
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (34,38,'/data/uploads/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (35,40,'main.css');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (36,41,'25');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (37,44,'abou-haider-dev.loris.ca');
```
I don't think this is right?

ridz1208 replied:
GRRRR

ridz1208 commented on test/unittests/VisitTest.php:
```diff
@@ -53,37 +53,150 @@ protected function setUp()
     function testAllVisit()
     {
         $visits = $this->_visitController->getAllVisits();
         $this->assertEquals($this->_listOfVisit, $visits, "the name of the visit does not match value in DB");
```
@driusan this assertequals seems to be comparing arrays in an ordered fashion which forced me to add an `order by` clause to the visitController for the array returned by the controller to match the one here.

I'm not sure thats normal behaviour...

Same goes for the one remaining error on the PR for function `testVisitsProjects`  the ordering seems to be the issue

driusan commented on docs/config/config.xml:
```diff
@@ -37,18 +37,18 @@
             <structure>
                 <seq type="siteAbbrev"/>
-                <seq type="alphanumeric" minLength="4"/> <!-- Ex: AAA1-->
+                <seq type="numeric" length="3" min='100' max='999'/> <!-- Ex: AAA1-->
```
Why is this changed in the default config?

driusan commented on docs/config/config.xml:
```diff
@@ -61,27 +61,48 @@
+        <!-- certification module -->
         <Certification>
-            <EnableCertification>0</EnableCertification>
+            <EnableCertification>1</EnableCertification>
```
I don't think this should be the default for LORIS, it makes it harder to set things up on a new install

ridz1208 commented:
@driusan makes sense for the config.xml. I'll try to find a solution where we can have a config for raisinbread alone and ideally not duplicate the default one.

as for the sourcing... not sure we need anything special. I'll maybe add a command to `cat` everything together into one big SQL

ridz1208 commented:
@driusan please re-review

ridz1208 commented:
@driusan I added some documentation. it could be improved later on if need be

driusan commented:
@ridz1208 Where's the documentation? (The PR changes 183 files, so it's not obvious in the diff..)

ridz1208 commented:
@driusan its the only .md file

https://github.com/aces/Loris/pull/4120/files#diff-891d4b84b8f410eef353c238c8045ab6

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+*This directory contains all the files necessary to spawn an instance of the Raisin Bread demonstration dataset.*
```
can you format the paragraphs in this document with `fmt` to make it readable?

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+
+***Note:** The following instructions assume that the user has already installed a functional version of LORIS. It's important to note that the steps below will clear all data from the currently active database and replace it with the RaisinBread data.*
+
+***Assumption:** The commands below assume that the MySQL configuration files have exist and contain a default database, user, host and password. In the event that that is not the case, replace all `mysql` commands below by the necessary values `mysql -u user -p -h host database_name`*
```
"have exist"?

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+***Assumption:** The commands below assume that the MySQL configuration files have exist and contain a default database, user, host and password. In the event that that is not the case, replace all `mysql` commands below by the necessary values `mysql -u user -p -h host database_name`*
+
+##### Sourcing SQL
+In order to be able to use the RaisinBread dataset, the LORIS SQL schema needs to be sourced, followed by the different instrument schemas and finally the actual RB data. The commands below assume that the current working directory is the main loris root directory.
```
s/loris/LORIS

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+### Modifying RB
+The RaisinBread database should be handled like any other project. The data should never be modified directly in the SQL files, the database should be sourced, altered and re-exported into SQL files.
+
+When possible, data modifications should be done directly from the LORIS front-end or using the loris tools. When modifications need to be done directly in SQL, it is important to only add clean data with proper realistic logical values where possible.
```
"loris" again.

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+### Modifying RB
+The RaisinBread database should be handled like any other project. The data should never be modified directly in the SQL files, the database should be sourced, altered and re-exported into SQL files.
+
+When possible, data modifications should be done directly from the LORIS front-end or using the loris tools. When modifications need to be done directly in SQL, it is important to only add clean data with proper realistic logical values where possible.
```
I don't know what "proper realistic logical values" are.

ridz1208 replied:
noone does

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,56 @@
+The RaisinBread data should be updated with every code change impacting the database. To do so, the user needs to source the database before the changes, modify the database and then export the data to be submitted back to LORIS in the same pull request as the code changes.
+
+##### Contributing back into LORIS
+In order to submit changes made to RaisinBread back to LORIS, an export tool was created in the `tools/exporters/` directory. The `DB_dump_table_data.php` PHP script executes an advanced `mysqldump` command that generates a single SQL file per SQL table in the `raisinbread/RB_files/` directory. Each file contains foreign keys disabling and enabling, table truncation to clear all existing data and ...
```
What makes it "advanced"?

ridz1208 replied:
it has a bunch of parameters, replace with "complex" ?

driusan replied:
I don't think any extra adjective at all is required.

ridz1208 commented:
@driusan done !!

driusan commented on raisinbread/README.md:
````diff
@@ -0,0 +1,96 @@
+the current working directory is the main LORIS root directory.
+
+``` cat SQL/0000-00-00-schema.sql \
+    SQL/0000-00-01-Permission.sql \ SQL/0000-00-02-Menus.sql \
````
code block formatting is messed up.

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,96 @@
+using the RaisinBread dataset some configurations are necessary.
+
+1. Copy the `raisinbread/config.config.xml` file into `project/config.xml`
+2. Input the correct `<database>` information in the `project/config.xml`
```
list is missing line breaks

driusan commented on raisinbread/README.md:
```diff
@@ -0,0 +1,96 @@
+`raisinbread/instruments/` instrument PHP and LINST files to the
+`projects/instruments/` directory
+
+### Modifying RB The RaisinBread database should be handled like any other
```
This (and other headings) seems to be missing a line break now.

ridz1208 commented:
@driusan should look better now

johnsaigle commented:
Also the `bids*` tables need to be dropped before the MRI tables. So move lines 116-119 above line 96.

ridz1208 commented:
@johnsaigle these can be fixed seperately, it has nothing to do with raisinbread its the default schemas

ridz1208 commented:
@johnsaigle I made the changes separately to avoid more files changed 
https://github.com/aces/Loris/pull/4459/files

christinerogers commented:
Tagging "Document at Release" because moving bmi.linst and other files from docs/instruments also requires updating the Loris Setup Guide (wiki/markdown) for Behavioural instruments.
